include/lbann/models/model.hpp

```diff
@@ -329,39 +327,58 @@ class model {
   /** Check if the model execution mode is valid. */
   virtual bool is_execution_mode_valid(execution_mode mode) const;
 
-  /** Reorder layers. */
-  virtual void permute_layers(const std::vector<int>& permutation);
+  /** Reorder layer list with a gather.
+   *
+   *  The new layer list is the same length as @c gather_indices and
+   *  its entries are given by
+   *  @f[ \text{new\_list}[i] = \text{old\_list}[\text{gather\_indices}[i]] @f]
+   *
+   *  Since entries in the layer list must be unique, this will fail
+   *  if @c gather_indices has any repeated entries.
+   */
+  void reorder_layers(const std::vector<El::Int>& gather_indices);
 
   /** Remap pointers.
+   *
    *  Layer and weights pointers are remapped using the provided
    *  maps. If a pointer is not a key in the corresponding map, the
    *  pointer is not changed.
    */
-  virtual void remap_pointers(const std::unordered_map<Layer *,Layer *>& layer_map,
-                              const std::unordered_map<weights *,weights *>& weights_map);
-
-  /** In case that a layer is frozen, also freeze layers that precede it if that
-   *  makes senses for the particular model, such as sequential or siamese.
-   *  For othe models, users can manually control the behaivor by indicating
-   *  whether to freeze each layer in the model description prototext.
+  virtual void remap_pointers(const std::unordered_map<Layer*,Layer*>& layer_map,
+                              const std::unordered_map<weights*,weights*>& weights_map);
+
+  /** @brief
+   *
+   *  In case that a layer is frozen, also freeze layers that precede
+   *  it if that makes senses for the particular model, such as
+   *  sequential or siamese.  For othe models, users can manually
+   *  control the behaivor by indicating whether to freeze each layer
+   *  in the model description prototext.
+   *
+   *  For general DAG models, users need to manually specify each
+   *  layer to freeze in the model description prototext.
    */
-  virtual void freeze_layers_under_frozen_surface();
```

Comment:
This function has an implementation in models/model.cpp, but it's overridden by an empty function in models/directed_acyclic_graph.hpp. I've gone ahead and removed the unused implementation. In the future, we may want to think about how to properly freeze layers in DAGs.